libs/contrib/Data/List/Reverse.idr

```diff
@@ -0,0 +1,93 @@
+||| Properties of the reverse function.
+module Data.List.Reverse
+
+import Data.List
+import Data.List.Equalities
+
+%default total
+
+export
+reverseOntoAcc : (xs, ys, zs : List a) ->
+  reverseOnto (ys ++ zs) xs = (reverseOnto ys xs) ++ zs
+reverseOntoAcc [] _ _ = Refl
+reverseOntoAcc (x :: xs) (ys) (zs) = reverseOntoAcc xs (x :: ys) zs
+
+||| Serves as a specification for reverseOnto.
+export
+reverseOntoSpec : (xs, ys : List a) -> reverseOnto xs ys = reverse ys ++ xs
+reverseOntoSpec xs ys = reverseOntoAcc ys [] xs
+
+||| The reverse of an empty list is an empty list.  Together with reverseCons,
+||| serves as a specification for reverse.
+export
+reverseNil : reverse [] = []
+reverseNil = Refl
+
+||| The reverse of a cons is the reverse of the tail followed by the head.
+||| Together with reverseNil serves as a specification for reverse.
+export
+reverseCons : (x : a) -> (xs : List a) -> reverse (x::xs) = reverse xs ++ [x]
+reverseCons x xs = reverseOntoSpec [x] xs
+
+||| Reversing an append is appending reversals backwards.
+export
+reverseAppend : (xs, ys : List a) ->
+  reverse (xs ++ ys) = reverse ys ++ reverse xs
+reverseAppend [] ys = sym (appendNilRightNeutral (reverse ys))
+reverseAppend (x :: xs) ys =
+  rewrite reverseCons x (xs ++ ys) in
+    rewrite reverseAppend xs ys in
+      rewrite reverseCons x xs in
+        sym $ appendAssociative (reverse ys) (reverse xs) [x]
+
+||| A recursive definition of reverse.
+export
+reverseRec : List a -> List a
```

Comment:
I wonder whether it's worth calling this one `slowReverse`?
Or to declare it as a `0 reverse : ...` to make sure it never
shows up in people's runtime code?

Reply:
I definitely like the idea of using `0` to mark specification functions.

On the other hand, maybe you sometimes just want to avoid unnecessary proofs when a simple/slow reverse would do.

Reply:
Oh by the way, if this function is a spec, shouldn't it be `public export`?

Reply:
Thanks. Updated PR.